src/error_decorator.cpp

```diff
@@ -0,0 +1,43 @@
+// quick-lint-js finds bugs in JavaScript programs.
+// Copyright (C) 2020  Matthew Glazar
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#include <quick-lint-js/error_decorator.h>
+#include <variant>
+
+namespace quick_lint_js {
+#define QLJS_ERROR_TYPE(name, struct_body, format_call)   \
+void error_decorator::report(name e) {                    \
+  error_reporter::error = true;                           \
+  std::visit([&](auto r) { r.report(e); }, reporter_);    \
```

Comment:
Blocking: Do not copy the underlying reporter. Use the existing reporter object:
```suggestion
  std::visit([&](auto &r) { r.report(e); }, reporter_);    \
```

Reply:
I totally modified this snippet so `std::visit` is not used anymore.